### ISSUE TITLE
[Win32] Different results from Time::HiRes::gettimeofday under the debugger

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7430]
Migrated from [rt.perl.org#30755](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=30755) (status was 'resolved')

Searchable as RT30755$

**Comment by p5pRT** [comment]
### From nigelsandever@btconnect.com

#### Created by nigelsandever@btconnect.com



Using Time&#8203;::HiRes&#8203;::gettimeofday\(\) in the debugger produces different\, much lower 
resolution\, results than it does from within a program\.

perl \-MTime&#8203;::HiRes=gettimeofday \-le" print scalar gettimeofday\(\) for 1 \.\. 10"
1090308821\.39091
1090308821\.39107
1090308821\.39114
1090308821\.39121
1090308821\.39127
1090308821\.39134
1090308821\.3914
1090308821\.39146
1090308821\.39152
1090308821\.39159

perl \-MTime&#8203;::HiRes=gettimeofday \-lde" print scalar gettimeofday\(\) for 1 \.\. 100"

Loading DB routines from perl5db\.pl version 1\.25
Editor support available\.

Enter h or \`h h' for help\, or \`perldoc perldebug' for more help\.

main&#8203;::\(\-e&#8203;:1\)&#8203;:    print scalar gettimeofday\(\) for 1 \.\. 100
&nbsp;  DB\<1> s
main&#8203;::\(\-e&#8203;:1\)&#8203;:    print scalar gettimeofday\(\) for 1 \.\. 100
&nbsp;  DB\<1> s
&nbsp;  \[SNIP\]
1090309302\.28125
1090309302\.28125
1090309302\.28125
1090309302\.29688
1090309302\.29688
1090309302\.29688
1090309302\.29688
1090309302\.29688
1090309302\.29688
1090309302\.29688
1090309302\.29688
1090309302\.29688
1090309302\.29688
1090309302\.3125
1090309302\.3125
1090309302\.3125
1090309302\.3125
1090309302\.3125
1090309302\.3125
1090309302\.3125
1090309302\.3125
1090309302\.32813
1090309302\.32813
1090309302\.32813
&nbsp;  \[SNIP\]
Debugged program terminated\.  Use q to quit or R to restart\,
&nbsp;  use O inhibit\_exit to avoid stopping after program termination\,
&nbsp;  h q\, h R or h O to get additional info\.
&nbsp;  DB\<1> q

See http&#8203;://perlmonks\.com/index\.pl?node\_id=375719 for my best guess as to the 
cause\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.4:

Configured by ActiveState at Tue Jun  1 11:52:09 2004.

Summary of my perl5 (revision 5 version 8 subversion 4) configuration:
  Platform:
    osname=MSWin32, osvers=4.0, archname=MSWin32-x86-multi-thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=undef use5005threads=undef useithreads=define 
usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -Gf -W3 -MD -Zi -DNDEBUG -O1 -DWIN32 -D_CONSOLE -
DNO_STRICT -DHAVE_DES_FCRYPT  -DNO_HASH_SEED -DPERL_IMPLICIT_CONTEXT -
DPERL_IMPLICIT_SYS -DUSE_PERLIO -DPERL_MSVCRT_READFIX',
    optimize='-MD -Zi -DNDEBUG -O1',
    cppflags='-DWIN32'
    ccversion='', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64', 
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf  -libpath:"c:
\Perl\lib\CORE"  -machine:x86'
    libpth=C:\PROGRA~1\MICROS~3\VC98\lib
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib 
uuid.lib wsock32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib 
msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib 
uuid.lib wsock32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib 
msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl58.lib
    gnulibc_version='undef'
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -debug -opt:ref,icf  -
libpath:"c:\Perl\lib\CORE"  -machine:x86'

Locally applied patches:
    ACTIVEPERL_LOCAL_PATCHES_ENTRY
    22751 Update to Test.pm 1.25
    21540 Fix backward-compatibility issues in if.pm


@INC for perl v5.8.4:
    c:/Perl/lib
    c:/Perl/site/lib
    .


Environment for perl v5.8.4:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=c:\Perl\bin;C:\Perl\bin\;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS
\System32\UnixTools;c:\perl584\bin;c:\perl561\bin;c:\perl581\bin;c:\perl561
\bin;c:\cl;
    PERL_BADLANG (unset)
    SHELL (unset)




```
</details>

**Comment by p5pRT** [comment]
### From @steve-m-hay

nigelsandever@&#8203;btconnect\.com \(via RT\) wrote&#8203;:

>Using Time&#8203;::HiRes&#8203;::gettimeofday\(\) in the debugger produces different\, much lower 
>resolution\, results than it does from within a program\.
>
I think the problem is the code within the "if \(diff \< \-MAX\_DIFF || diff 
 > MAX\_DIFF\) \{" section which is intended to recalibrate to the system 
time if we have drifted too far from it\.

It is triggered when running under the debugger \(relating to the delay 
between starting the debugger and running the code in \-e "\.\.\."\)\, but the 
reset doesn't appear to be functioning correctly\.  I couldn't help 
thinking that this doesn't look right&#8203;:

&nbsp;    QueryPerformanceCounter\(\(LARGE\_INTEGER\*\)&ticks\);
&nbsp;    ticks \-= MY\_CXT\.base\_ticks;
&nbsp;    \.\.\.
&nbsp;    if \(resetting\)
&nbsp;        MY\_CXT\.base\_ticks = ticks;

Having taken base\_ticks off ticks\, shouldn't it be added back to ticks 
before ticks is assigned to base\_ticks\, i\.e\.

&nbsp;    if \(resetting\)
&nbsp;        MY\_CXT\.base\_ticks = ticks \+ MY\_CXT\.base\_ticks;

Also\, the original initialisation section sets up 
MY\_CXT\.base\_systime\_as\_filetime and ft&#8203;:

&nbsp;    GetSystemTimeAsFileTime\(&MY\_CXT\.base\_systime\_as\_filetime\.ft\_val\);
&nbsp;    ft\.ft\_i64 = MY\_CXT\.base\_systime\_as\_filetime\.ft\_i64;

but only the latter gets reset when we have drifted&#8203;:

&nbsp;    GetSystemTimeAsFileTime\(&filtim\.ft\_val\);
&nbsp;    \.\.\.
&nbsp;    ft\.ft\_i64 = filtim\.ft\_i64;

Doesn't MY\_CXT\.base\_systime\_as\_filetime need resetting too?

The attached patch \(against Time\-HiRes\-1\.59\) makes these changes and \(a\) 
fixes the reported problem and \(b\) still passes the test suite\.

\- Steve




---
Radan Computational Ltd\.

The information contained in this message and any files transmitted with it are confidential and intended for the addressee\(s\) only\.  If you have received this message in error or there are any problems\, please notify the sender immediately\.  The unauthorized use\, disclosure\, copying or alteration of this message is strictly forbidden\.  Note that any views or opinions presented in this email are solely those of the author and do not necessarily represent those of Radan Computational Ltd\.  The recipient\(s\) of this message should check it and any attached files for viruses&#8203;: Radan Computational will accept no liability for any damage caused by any virus transmitted by this email\.

**Comment by p5pRT** [comment]
### From @steve-m-hay

<details><summary>Inline Patch</summary>

```diff
--- HiRes.xs.orig	2004-04-08 20:55:42.000000000 +0100
+++ HiRes.xs	2004-07-24 23:34:51.000000000 +0100
@@ -197,8 +197,6 @@
 
     if (MY_CXT.run_count++) {
 	__int64 diff;
-	FT_t filtim;
-	GetSystemTimeAsFileTime(&filtim.ft_val);
         QueryPerformanceCounter((LARGE_INTEGER*)&ticks);
         ticks -= MY_CXT.base_ticks;
         ft.ft_i64 = MY_CXT.base_systime_as_filetime.ft_i64
@@ -206,8 +204,9 @@
                     +(Const64(10000000) * (ticks % MY_CXT.tick_frequency)) / MY_CXT.tick_frequency;
 	diff = ft.ft_i64 - MY_CXT.base_systime_as_filetime.ft_i64;
 	if (diff < -MAX_DIFF || diff > MAX_DIFF) {
-	     MY_CXT.base_ticks = ticks;
-	     ft.ft_i64 = filtim.ft_i64;
+	     MY_CXT.base_ticks += ticks;
+	     GetSystemTimeAsFileTime(&MY_CXT.base_systime_as_filetime.ft_val);
+	     ft.ft_i64 = MY_CXT.base_systime_as_filetime.ft_i64;
 	}
     }
     else {
```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @steve-m-hay

Fixed in Time&#8203;::HiRes 1\.60\, merged into bleadperl as change \#23232\.  Will
be in perl 5\.8\.6\.

**Comment by p5pRT** [comment]
@steve-m-hay - Status changed from 'open' to 'resolved'
